source/common/network/apple_dns_impl.cc

```diff
@@ -402,6 +209,69 @@ AppleDnsResolverImpl::PendingResolution::dnsServiceGetAddrInfo(DnsLookupFamily d
       this);
 }
 
+void AppleDnsResolverImpl::PendingResolution::onDNSServiceGetAddrInfoReply(
+    DNSServiceFlags flags, uint32_t interface_index, DNSServiceErrorType error_code,
+    const char* hostname, const struct sockaddr* address, uint32_t ttl) {
+  ENVOY_LOG(debug,
+            "DNS for {} resolved with: flags={}[MoreComing={}, Add={}], interface_index={}, "
+            "error_code={}, hostname={}",
+            dns_name_, flags, flags & kDNSServiceFlagsMoreComing ? "yes" : "no",
+            flags & kDNSServiceFlagsAdd ? "yes" : "no", interface_index, error_code, hostname);
+
+  // Generic error handling.
+  if (error_code != kDNSServiceErr_NoError) {
+    // TODO(junr03): consider creating stats for known error types (timeout, refused connection,
+    // etc.). Currently a bit challenging because there is no scope access wired through. Current
+    // query gets a failure status
+
+    pending_cb_.status_ = ResolutionStatus::Failure;
+    pending_cb_.responses_.clear();
+
+    finishResolve();
+    // Note: Nothing can follow this call to flushPendingQueries due to deletion of this
+    // object upon resolution.
+    return;
+  }
+
+  // Only add this address to the list if kDNSServiceFlagsAdd is set. Callback targets are only
+  // additive.
+  if (flags & kDNSServiceFlagsAdd) {
+    ASSERT(address, "invalid to add null address");
+    auto dns_response = buildDnsResponse(address, ttl);
+    ENVOY_LOG(debug, "Address to add address={}, ttl={}",
+              dns_response.address_->ip()->addressAsString(), ttl);
+    pending_cb_.responses_.push_back(dns_response);
+  }
```

Comment:
To matt's comment: What happens on the else here? aka `flags & kDNSServiceFlagsAdd == 0` which implies "removal"?

Reply:
> If a successfully-registered name later suffers a name conflict or similar problem and has to be deregistered, the callback will be invoked with the kDNSServiceFlagsAdd flag not set.

Reply:
@mattklein123 consolidating your [question](https://github.com/envoyproxy/envoy/pull/17226#discussion_r664664756) here to keep everything in one place.

The current resolver API is purely additive and up until now we have let the consumer of the resolver (e.g, logical dns cluster, dns cache, etc.) deal with the result from the query as they see fit. So for instance the logical DNS cluster will take the [first address](https://github.com/envoyproxy/envoy/blob/main/source/common/upstream/logical_dns_cluster.cc#L127) in the response, and replace the existing value [if it is different (or one doesn't exist)](https://github.com/envoyproxy/envoy/blob/main/source/common/upstream/logical_dns_cluster.cc#L146). So removal is implicit and is based on the consumer's behavior.

@mattklein123 I don't understand what you mean by  

> Will it always "add" things that have already been added?

do you mind expanding on that?

Reply:
What I mean is the apple API seems to internally support add/remove. Do we have to handle removal? Or because we are using a new connection for each query it doesn't matter? Can you add more comments on that?

Reply:
Oh, I think I see what you mean now @mattklein123. Given that the API handles removal, it seems to internally track what has happened with a particular DNS name in the past and keep state. So the concern is that potentially two independent queries from consumers of the resolver would interfere with each other. For instance, potentially a subsequent query would only receive a "remove" response and end up with no addresses, because the resolution doesn't always "add" things that have already been added, as you mentioned originally. 

Interesting. Let me read through the API docs again.